### ISSUE TITLE
Computing `<MyStruct as std::ptr::Pointee>::Metadata` fails when trailing field has a projection

### ISSUE THREAD

**Issue by Zalathar** [rust-lang/rust#92577]
### Code

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8e31db4a857f6bcdbf287b8d9745ca22

```Rust
use std::collections::HashMap;

trait HasType {
    type Type;
}

impl HasType for () {
    type Type = ();
}

pub struct MyStruct {
    _field: <() as HasType>::Type,
}

pub fn do_insert(map: &mut HashMap<*const MyStruct, ()>, key: *const MyStruct) {
    map.insert(key, ());
}
```

### Meta

(Build fails with an ICE on current stable/beta/nightly on the playground.)

### Error output

```
   Compiling playground v0.0.1 (/playground)
thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', compiler/rustc_monomorphize/src/collector.rs:894:93
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.57.0 (f1edd0429 2021-11-29) running on x86_64-unknown-linux-gnu

note: compiler flags: -C embed-bitcode=no -C codegen-units=1 -C debuginfo=2 --crate-type lib

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [collect_and_partition_mono_items] collect_and_partition_mono_items
#1 [exported_symbols] exported_symbols
end of query stack
error: could not compile `playground`
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
   Compiling playground v0.0.1 (/playground)
thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', compiler/rustc_monomorphize/src/collector.rs:894:93
stack backtrace:
   0: rust_begin_unwind
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panicking.rs:100:14
   2: core::panicking::panic
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panicking.rs:50:5
   3: <rustc_monomorphize::collector::MirNeighborCollector as rustc_middle::mir::visit::Visitor>::visit_terminator
   4: rustc_monomorphize::collector::collect_neighbours
   5: rustc_monomorphize::collector::collect_items_rec
   6: rustc_monomorphize::collector::collect_items_rec
   7: rustc_monomorphize::collector::collect_items_rec
   8: rustc_monomorphize::collector::collect_items_rec
   9: rustc_monomorphize::collector::collect_items_rec
  10: rustc_session::utils::<impl rustc_session::session::Session>::time
  11: rustc_monomorphize::collector::collect_crate_mono_items
  12: rustc_monomorphize::partitioning::collect_and_partition_mono_items
  13: rustc_query_system::query::plumbing::try_execute_query
  14: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::collect_and_partition_mono_items
  15: rustc_codegen_ssa::back::symbol_export::exported_symbols_provider_local
  16: rustc_query_system::query::plumbing::try_execute_query
  17: rustc_query_system::query::plumbing::get_query
  18: rustc_metadata::rmeta::encoder::EncodeContext::encode_crate_root
  19: rustc_metadata::rmeta::encoder::encode_metadata_impl
  20: rustc_data_structures::sync::join
  21: rustc_metadata::rmeta::encoder::encode_metadata
  22: rustc_interface::queries::Queries::ongoing_codegen
  23: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  24: rustc_span::with_source_map
  25: scoped_tls::ScopedKey<T>::set
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.57.0 (f1edd0429 2021-11-29) running on x86_64-unknown-linux-gnu

note: compiler flags: -C embed-bitcode=no -C codegen-units=1 -C debuginfo=2 --crate-type lib

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [collect_and_partition_mono_items] collect_and_partition_mono_items
#1 [exported_symbols] exported_symbols
end of query stack
error: could not compile `playground`
```

</p>
</details>

**Comment by Zalathar** [comment]
Same error message as #90738, though it's unclear whether that issue is related.

**Comment by Zalathar** [comment]
[According to Godbolt this worked in 1.51.0, and started failing in 1.52.0.](https://rust.godbolt.org/z/T3cMcGMTP)

**Comment by Aaron1011** [comment]
This appears to be fixed in the latest nightly.

**Comment by Aaron1011** [comment]
Minimized:

```rust
#![feature(ptr_metadata)]

trait HasType {
    type Type;
}

impl HasType for () {
    type Type = ();
}

pub struct MyStruct {
    _field: <() as HasType>::Type,
}

pub fn bar() {
    let val: <MyStruct as std::ptr::Pointee>::Metadata;
}
```

The `Pointee` trait gets used by the `Hash` impl for `*const MyStruct`, which leads to the bad monomorphization:

https://github.com/rust-lang/rust/blob/181e91567c9f347e055b33b1d7e9894f769aafe3/library/core/src/hash/mod.rs#L791-L799

The issue is caused by how we handle projections when computing the `Metadata` type. We don't normalize the 'tail' type: https://github.com/rust-lang/rust/blob/181e91567c9f347e055b33b1d7e9894f769aafe3/compiler/rustc_trait_selection/src/traits/project.rs#L1403-L1404

which causes us to not create a candidate for the projection type in the tail of `MyField` (which is `<() as HasType>::Type`):

https://github.com/rust-lang/rust/blob/181e91567c9f347e055b33b1d7e9894f769aafe3/compiler/rustc_trait_selection/src/traits/project.rs#L1432-L1438

cc @SimonSapin - we need to be normalizing *somwhere*, but I'm not exactly sure where the correct place is.

**Comment by SimonSapin** [comment]
I’m sorry but I don’t think I can help much. I don’t even know what normalizing means precisely. As I mentioned in https://github.com/rust-lang/rust/pull/81172

> For this I mostly imitated what the compiler was already doing for the `DiscriminantKind` trait. I’m very unfamiliar with compiler internals

So the only hint I can offer is: does the code for `DiscriminantKind` do something differently? Or is there a reason it’s not affected by this problem?

**Comment by Aaron1011** [comment]
cc @rust-lang/wg-traits - I'm not quite sure what the proper normalization behavior is. We need to be able to perform this projection during typecheck (e.g. `let a: <MyType as Pointee>::Metadata = foo();`), so I think we need to try to faillibly normalize the trailing field type here?

**Comment by jackh726** [comment]
Why are we *not* normalizing? Because we don't want to replace with an inference var?

**Comment by compiler-errors** [comment]
@Aaron1011, so I put up a fix for this in #92248. Someone mind reviewing that PR? It's gone some time without being touched.

> we need to be normalizing somewhere, but I'm not exactly sure where the correct place is.

Is it really not sufficient to just normalize during selection? My PR seems to successfully compile both examples linked in this issue.
